### DISCUSSION TITLE
Common-Lisp: defpackage vs in-package for context

### DISCUSSION THREAD

**Discussion by iicurtis** [Olical/conjure#557] [General]
When talking with the REPL, common-lisp needs to define a context to operate in. This is currently done by looking for the most recent `(defpackage __context__)`. Does it make sense to instead look for `(in-package __context__)`? Or maybe either? (Not sure if that's possible with lua patterns)

**Comment by Olical** [comment]
We should be able to do that, yeah. It means adapting https://github.com/Olical/conjure/blob/f50d4db821489f66ff93afe58155eef42c273a5d/fnl/conjure/client/common-lisp/swank.fnl#L15

The client API also supports a `context` function instead of a `context-pattern` string, so if we want to we could drop that pattern and export a function that checks for both, the context-pattern string is really just shorthand, it's more flexible than that.

So yeah, we could quite easily check for both in a particular order and make some decision about what the current context for the current document is. Maybe even base it on where in the file your cursor is right now for more contextual evals?

**Reply by Olical** [reply]
Pushed a change to develop noted in #561 and https://github.com/Olical/conjure/commit/20df0e0c4dc41654cf7276801e0aa2f4931cce6f

Let me know what you think!

**Comment by russtoku** [comment]
I don't know much about Common Lisp but some research provided some help:

- In the _How to load an existing project_ section of [The Common Lisp Cookbook – Getting started with Common Lisp](https://lispcookbook.github.io/cl-cookbook/getting-started.html), it suggest that `(in-package __context__)` is what we should look for because forms evaluated after that should be in that package.

- In [Chapter 20. Special Operators](https://gigamonkeys.com/book/the-special-operators) of [Practical Common Lisp](https://gigamonkeys.com/book/), Peter Seibel says:

  > evaluating an IN-PACKAGE form changes the value of *PACKAGE*, which will affect the way subsequent forms are read.

- In the _Appendix: Packages_ section of [On Lisp](https://www.paulgraham.com/onlisp.html) by Paul Graham, he says:

  > By calling in-package we can switch to a new package, creating it if necessary:
  > ```
  > > (in-package ’mine :use ’common-lisp) #<Package "MINE" 63390E>
  > ```
  > At this point there should be eerie music, because we are in a different world: foo here is not what it used to be.

I've also tried out https://github.com/monkoose/nvlime to see how the interaction with a Common Lisp REPL works. It does switch packages when it evaluates an `in-package` form. It also has a key mapping that lets you change the current package.

**Comment by Olical** [comment]
Closing this one because I _think_ I addressed it. Please do open an issue if not!
